document use of gds-cli to deploy terraform (#2590)
alphagov/govuk-developer-docs: source/manual/deploying-terraform.html.md

You can find which class of user you are [infra-security project in govuk-aws-data](https://github.com/alphagov/govuk-aws-data/tree/master/data/infra-security).

### 2. Install `gds-cli`

> [`gds-cli`](/manual/access-aws-console.html) is the preferred way of obtaining

Reviewer: This now redirects to `/manual/get-started.html`, so should be updated to that.

Except I also find this link quite surprising - I'd expect `gds-cli` to link to the gds-cli repo.

Perhaps

```suggestion
[`gds-cli`](https://github.com/alphagov/gds-cli) is the [preferred way](/manual/get-started.html) of obtaining
```

AWS credentials.

> As of version `v2.15.0` of [`gds-cli`](/manual/access-aws-console.html),

Reviewer: You link to the same place as the line above - I'd suggest just removing this link altogether.

you can use it to deploy terraform via Jenkins.

> ### 3. Get Github Credentials

Reviewer: ```suggestion
### 3. Get GitHub Credentials
```

You need to obtain your Github credentials by creating a read-only GitHub personal access token. This [GitHub personal access token](https://github.com/settings/tokens) should be
created with the `read:org` scope only.

> Take care to store and handle the token securely. If you accidentally share your token,
  [revoke it immediately](https://github.com/settings/tokens) and follow the
  [instructions for reporting a potential data security incident][security-incidents].

## Deploying Terraform

**Always `plan` first, check that the output is what you expect, then `apply`.**

There are 2 ways of deploying terraform:

> 1. [`gds-cli`](/manual/access-aws-console.html) as of version `v2.15.0`

Reviewer: I'd remove the link here too, as I don't think it adds a lot.

Author: I added a ref link at the bottom so that the link definition is done once only but but many times

2. `deploy.rb` script in [govuk-aws][deploy-rb] repository

> ### 1. [`gds-cli`](/manual/access-aws-console.html) as of version `v2.15.0`

Reviewer: I don't think we need to repeat the version and link throughout - this would do:

```suggestion
### 1. gds-cli
```

> To deploy terraform using [`gds-cli`](/manual/access-aws-console.html), you should run:

Reviewer: And remove the link here.

Author: I added a ref link at the bottom so that the link definition is done once only but but many times

```sh
GITHUB_USERNAME=<github_username> GITHUB_TOKEN=<github_token> \
gds govuk terraform -e <environment> -p <project> -s <stack> -a <action> -r <aws_role>
```

> where:

Reviewer: ```suggestion
Where:
```

1. `<github_username>` is the name of your Github account

> 1. `<github_token>` is the Github token that you created as described [above](#3-get-github-credentials)

Reviewer: `GitHub` rather than `github` for the two points above. (Nitpicky, I know!)

Author: done

1. `<environment>` is the govuk environment you want to deploy to. E.g. `integration`,`staging`
1. `<project>` is the terraform project that you want to deploy. E.g. `app-gatling`
1. `<stack>` is the govuk stack you want to deploy to. E.g. `blue`, `govuk`
1. `<action>` is the terraform action you want to perform. E.g. `plan`, `apply`
1. `<aws_role>` is the govuk aws role you want to use for terraforming. E.g. `govuk-integration-admin`

After you deploy, you can visit the [`ci-deploy` Jenkins job][ci-deploy-jenkins] to see the job running or queued.

> ### 2. `deploy.rb` script in [govuk-aws][deploy-rb] repository

Reviewer: Heading could be simplified IMO:

```suggestion
### 2. `deploy.rb` script in govuk-aws
```

The Ruby script [`tools/deploy.rb`][deploy-rb] in the `govuk-aws` repository takes care of requesting temporary
AWS credentials with an assumed role and queuing the deployment Jenkins job.

You can use it by running:

```sh
GITHUB_USERNAME=<your GitHub username> \
  GITHUB_TOKEN=<your GitHub personal access token> \
  gds aws <your role e.g. govuk-integration-admin> -- \
  ~/govuk/govuk-aws/tools/deploy.rb blue app-backend integration plan
```

You will need to change the arguments to the `deploy.rb` script. E.g.

* `app-backend` should be the name of the project you want to deploy
* `blue` is for `app-` projects, `govuk` is for `infra-` projects usually
* `integration` is the starting point, then `staging`, etc.

Once the script has run, visit the [`ci-deploy` Jenkins job][ci-deploy-jenkins] to see the job running or queued.

> 

Reviewer: Looks like an extra newline slipped in

```suggestion
```

### 2. Install `gds-cli`

[`gds-cli`](/manual/access-aws-console.html) is the preferred way of obtaining
AWS credentials.

As of version `v2.15.0` of [`gds-cli`](/manual/access-aws-console.html),
you can use it to deploy terraform via Jenkins.

### 3. Get Github Credentials

If you are **not** deploying terraform manually on the Jenkins boxes, you need to
obtain your Github credentials by creating a read-only GitHub personal access token.
This [GitHub personal access token](https://github.com/settings/tokens) should be
created with the `read:org` scope only.

> Take care to store and handle the token securely. If you accidentally share your token,
  [revoke it immediately](https://github.com/settings/tokens) and follow the
  [instructions for reporting a potential data security incident][security-incidents].

## Deploying Terraform

**Always `plan` first, check that the output is what you expect, then `apply`.**

There are 3 ways of deploying terraform:

1. [`gds-cli`](/manual/access-aws-console.html) as of version `v2.15.0`
2. `deploy.rb` script in [govuk-aws][deploy-rb] repository

> 3. Web interface of [Jenkins][ci-deploy-jenkins]

Reviewer 2: I think we can delete this one. The other two are safer. And in time, we can delete `2.` as well.

Author: cool, i can do that.
